db/functions/callbacks.go

```diff
@@ -0,0 +1,244 @@
+	if dbUser := dbc.User(); dbUser != nil {
+		d.user = &userCredentials{
+			Name:     dbUser.Name(),
+			Roles:    dbUser.RoleNames().AllKeys(),
+			Channels: dbUser.Channels().AllKeys(),
+		}
+	}
+	return nil
+}
+
+func (d *databaseDelegate) getCollection(collectionName string) (*db.DatabaseCollectionWithUser, error) {
+	return d.db.GetDatabaseCollectionWithUser(base.DefaultScope, collectionName)
+}
+
+// Temporarily gives the `db.Database` admin powers. Returns a fn that will revert the upgrade.
+// Must be called as `defer d.asAdmin()()`
+func (d *databaseDelegate) asAdmin() func() {
+	user := d.db.User()
+	d.db.SetUser(nil)
+	return func() { d.db.SetUser(user) } // <-- this is what will be run by `defer`
+}
+
+func (d *databaseDelegate) checkTimeout() error {
+	return db.CheckTimeout(d.ctx)
+}
+
+func (d *databaseDelegate) get(docID string, collectionName string, asAdmin bool) (doc map[string]any, err error) {
+	if asAdmin {
+		defer d.asAdmin()()
+	}
+	collection, err := d.getCollection(collectionName)
+	if err != nil {
+		return nil, err
+	}
+	rev, err := collection.GetRev(d.ctx, docID, "", false, nil)
+	if err != nil {
+		status, _ := base.ErrorAsHTTPStatus(err)
+		if status == http.StatusNotFound {
+			// Not-found is not an error; just return null.
+			return nil, nil
+		}
+		return nil, err
+	}
+	body, err := rev.Body()
+	if err != nil {
+		return nil, err
+	}
+	body["_id"] = docID
+	body["_rev"] = rev.RevID
+	return body, nil
+}
+
+func (d *databaseDelegate) save(body map[string]any, docID string, collectionName string, asAdmin bool) (bool, error) {
+	if asAdmin {
+		defer d.asAdmin()()
+	}
+	delete(body, "_id")
+	collection, err := d.getCollection(collectionName)
+	if err != nil {
+		return false, err
+	}
+	if _, found := body["_rev"]; found {
+		// If caller provided `_rev` property, use MVCC as normal:
+		if _, _, err := collection.Put(d.ctx, docID, body); err != nil {
+			if status, _ := base.ErrorAsHTTPStatus(err); status == http.StatusConflict {
+				err = nil // conflict: no error, but returns false
+			}
+			return false, err
+		}
+
+	} else {
+		// If caller didn't provide a `_rev` property, fall back to "last writer wins":
+		// get the current revision if any, and pass it to Put so that the save always succeeds.
+		for {
+			rev, err := collection.GetRev(d.ctx, docID, "", false, []string{})
+			if err != nil {
+				if status, _ := base.ErrorAsHTTPStatus(err); status != http.StatusNotFound {
+					return false, err
+				} else if del, found := body["_deleted"].(bool); found && del {
+					// Deleting nonexistent doc: success
+					return true, nil
+				}
+			}
+			if rev.RevID == "" {
+				delete(body, "_rev")
+			} else {
+				body["_rev"] = rev.RevID
+			}
+
+			_, _, err = collection.Put(d.ctx, docID, body)
+			if err == nil {
+				break // success!
+			} else if status, _ := base.ErrorAsHTTPStatus(err); status != http.StatusConflict {
+				return false, err
+			}
+			// on conflict (race condition), retry...
```

Comment:
is this  an important todo?

Reply:
It's not a todo, it's a description of what happens. This line is reached when there's a conflict, and since it's at the end of a `for` block it will go back up and retry.